src/AdvancedHMC.jl

```diff
@@ -146,7 +151,12 @@ function Hamiltonian(metric::AbstractMetric, ℓ; kwargs...)
     ℓπ = if cap === LogDensityProblems.LogDensityOrder{0}()
         # In this case ℓ does not support evaluation of the gradient of the log density function
         # We use ForwardDiff to compute the gradient
-        LogDensityProblemsAD.ADgradient(Val(:ForwardDiff), ℓ; kwargs...)
+        _logdensity = Base.Fix1(LogDensityProblems.logdensity, ℓ)
+        _logdensity_and_gradient = function (x)
+            prep = DI.prepare_gradient(_logdensity, adtype, x)
```

Comment:
What's the point of the preparation step in this function? Every time you'd evaluate logdensity + gradient for an input `x`, it would run preparation for `x` and then immediately discard it after the computation. If you just use `LogDensityProblemsAD.ADgradient` (see above), then users could provide a typical input `x` as keyword argument to optimize the AD computations of the hamiltonian (or otherwise they'd get the slower fallback).

Reply:
If preparation is not reused, then it is not beneficial. But as per my [comment above](https://github.com/TuringLang/AdvancedHMC.jl/pull/405#discussion_r2008703659), if it can be reused because we have access to a typical input, then it works basically the same as a `LogDensityProblemsAD.ADgradient` (except that Turing no longer depends on LogDensityProblemsAD` IIUC)

Reply:
Changing to DifferentiationInterface is just a step forward for the mutualized AD maintenance in Turing, and seems if we want to use DI to compute the gradient, we will have to use the preparation mechanism. As for the performance aspects, I will test the difference between using DifferentiationInterface and LogDensityProblemsAD, if no difference between those two options, we can safely change to DI then